File: src/wandb_mcp_server/mcp_tools/code_sandbox/pyodide_sandbox.ts
#!/usr/bin/env -S deno run --allow-net --allow-read --allow-write --allow-env

/**
 * Direct Pyodide sandbox implementation for Python code execution.
 * 
 * Based on Pydantic AI's mcp-run-python server:
 * https://github.com/pydantic/pydantic-ai/tree/main/mcp-run-python
 * 
 * We've adapted their approach of using Deno + Pyodide for sandboxed Python execution
 * with our own implementation that provides direct file system access and better
 * integration with our MCP server architecture.
 */

import { loadPyodide } from "npm:pyodide@0.26.4";

interface ExecutionRequest {
  type?: "execute" | "writeFile";
  code?: string;
  path?: string;
  content?: string;
  files?: { [path: string]: string };
  timeout?: number;
}

interface ExecutionResult {
  success: boolean;
  output: string;
  error: string | null;
  logs: string[];
}

class PyodideSandbox {
  private pyodide: any = null;
  private initialized = false;

  async initialize() {
    if (this.initialized) return;

    try {
      console.error("Initializing Pyodide...");
      this.pyodide = await loadPyodide({
        stdout: (text: string) => {
          // Redirect to stderr during initialization to avoid interfering with JSON responses
          console.error(`Pyodide stdout: ${text}`);
        },
        stderr: (text: string) => {
          // Redirect to stderr during initialization
          console.error(`Pyodide stderr: ${text}`);
        },
      });

      // Set up interrupt buffer for cancelling long-running code
      // This allows us to interrupt Python execution
      this.pyodide.setInterruptBuffer(new Uint8Array(new SharedArrayBuffer(4)));
      
      // Load commonly used packages - output will go to stderr
      console.error("Loading packages: numpy, pandas, matplotlib...");
      await this.pyodide.loadPackage(["numpy", "pandas", "matplotlib"]);
      
      this.initialized = true;
      console.error("Pyodide initialized successfully");
    } catch (error) {
      console.error("Failed to initialize Pyodide:", error);
      throw error;
    }
  }

  async execute(request: ExecutionRequest): Promise<ExecutionResult> {
    if (!this.initialized) {
      await this.initialize();
    }

    const result: ExecutionResult = {
      success: false,
      output: "",
      error: null,
      logs: [],
    };

    try {
      // Set up output capture BEFORE running any Python code
      
      // First, set up Python's sys.stdout and sys.stderr to capture output
      await this.pyodide.runPythonAsync(`
import sys
from io import StringIO

# Create StringIO objects to capture output
_stdout_buffer = StringIO()
_stderr_buffer = StringIO()

# Save original stdout/stderr
_original_stdout = sys.stdout
_original_stderr = sys.stderr

# Redirect stdout/stderr
sys.stdout = _stdout_buffer
sys.stderr = _stderr_buffer
      `);

      // Write any input files to the virtual filesystem
      if (request.files) {
        for (const [path, content] of Object.entries(request.files)) {
          try {
            // Ensure directory exists
            const dir = path.substring(0, path.lastIndexOf('/'));
            if (dir) {
              this.pyodide.FS.mkdirTree(dir);
            }
            
            // Write file
            this.pyodide.FS.writeFile(path, content);
            result.logs.push(`Wrote file: ${path}`);
          } catch (error) {
            result.logs.push(`Failed to write file ${path}: ${error}`);
          }
        }
      }

      // Execute the Python code with timeout
      const timeout = request.timeout || 30;
      
      // Set up timeout handling with interrupt
      let timeoutId: number | null = null;
      let interrupted = false;
      
      const timeoutPromise = new Promise((_, reject) => {
        timeoutId = setTimeout(() => {
          // Interrupt the Python execution
          try {
            this.pyodide.interruptExecution();
            interrupted = true;
          } catch (e) {
            console.error("Failed to interrupt execution:", e);
          }
          reject(new Error(`Execution timed out after ${timeout} seconds`));
        }, timeout * 1000);
      });

      try {
        // Race between execution and timeout
        const executionResult = await Promise.race([
          this.pyodide.runPythonAsync(request.code!),
          timeoutPromise
        ]);
        
        // Clear timeout if execution completed
        if (timeoutId !== null) {
          clearTimeout(timeoutId);
        }
        
        // Capture the output from the StringIO buffers
        const capturedOutput = await this.pyodide.runPythonAsync(`
# Get captured output
_stdout_output = _stdout_buffer.getvalue()
_stderr_output = _stderr_buffer.getvalue()

# Restore original stdout/stderr
sys.stdout = _original_stdout
sys.stderr = _original_stderr

# Clean up
_stdout_buffer.close()
_stderr_buffer.close()

# Return the captured output
(_stdout_output, _stderr_output)
        `);
        
        const [stdoutOutput, stderrOutput] = capturedOutput.toJs();
        
        result.success = true;
        result.output = stdoutOutput || "";
        
        // Add the return value if it's not None
        if (executionResult !== undefined && executionResult !== null && executionResult !== this.pyodide.globals.get('None')) {
          if (result.output && !result.output.endsWith('\n')) {
            result.output += "\n";
          }
          result.output += String(executionResult);
        }
        
        if (stderrOutput) {
          result.logs.push(stderrOutput);
        }
      } catch (error: any) {
        if (interrupted || error.message?.includes("timed out")) {
          // This is a timeout/interrupt error
          result.error = `Execution timed out after ${timeout} seconds`;
          result.success = false;
          
          // Try to capture any partial output
          try {
            const capturedOutput = await this.pyodide.runPythonAsync(`
# Get captured output even if there was a timeout
_stdout_output = _stdout_buffer.getvalue() if '_stdout_buffer' in locals() else ""
_stderr_output = _stderr_buffer.getvalue() if '_stderr_buffer' in locals() else ""

# Restore original stdout/stderr if they exist
if '_original_stdout' in locals():
    sys.stdout = _original_stdout
if '_original_stderr' in locals():
    sys.stderr = _original_stderr

# Clean up if buffers exist
if '_stdout_buffer' in locals():
    _stdout_buffer.close()
if '_stderr_buffer' in locals():
    _stderr_buffer.close()

(_stdout_output, _stderr_output)
            `);
            
            const [stdoutOutput, stderrOutput] = capturedOutput.toJs();
            if (stdoutOutput) {
              result.output = stdoutOutput;
            }
            if (stderrOutput) {
              result.logs.push(stderrOutput);
            }
          } catch {
            // Ignore errors in cleanup
          }
          
          return result;
        }
        
        // Regular error (not timeout)
        // Try to capture any output before the error
        try {
          const capturedOutput = await this.pyodide.runPythonAsync(`
# Get captured output even if there was an error
_stdout_output = _stdout_buffer.getvalue() if '_stdout_buffer' in locals() else ""
_stderr_output = _stderr_buffer.getvalue() if '_stderr_buffer' in locals() else ""

# Restore original stdout/stderr if they exist
if '_original_stdout' in locals():
    sys.stdout = _original_stdout
if '_original_stderr' in locals():
    sys.stderr = _original_stderr

# Clean up if buffers exist
if '_stdout_buffer' in locals():
    _stdout_buffer.close()
if '_stderr_buffer' in locals():
    _stderr_buffer.close()

(_stdout_output, _stderr_output)
          `);
          
          const [stdoutOutput, stderrOutput] = capturedOutput.toJs();
          if (stdoutOutput) {
            result.output = stdoutOutput;
          }
          if (stderrOutput) {
            result.logs.push(stderrOutput);
          }
        } catch {
          // Ignore errors in cleanup
        }
        
        // Handle Python errors specially
        if (error && error.constructor && error.constructor.name === "PythonError") {
          // Pyodide wraps Python exceptions - extract the actual error type
          const errorStr = error.toString();
          const errorMessage = error.message || errorStr;
          
          // Try to extract the Python error type from the traceback
          // Python tracebacks typically show the error type at the end
          const syntaxErrorMatch = errorMessage.match(/SyntaxError:|IndentationError:|NameError:|TypeError:|ValueError:|AttributeError:|KeyError:|IndexError:|ZeroDivisionError:/);
          if (syntaxErrorMatch) {
            result.error = errorMessage; // Include full traceback
          } else {
            // Fallback - at least mention it's a Python error
            result.error = `PythonError: ${errorMessage}`;
          }
        } else if (error && error.message) {
          result.error = error.message;
        } else {
          result.error = error.toString();
        }
      }

    } catch (error: any) {
      result.error = `Sandbox execution failed: ${error.toString()}`;
    }

    return result;
  }

  /**
   * Write a file directly to the Pyodide filesystem.
   * This is more efficient than executing Python code to write files.
   */
  async writeFile(path: string, content: string): Promise<void> {
    if (!this.initialized) {
      await this.initialize();
    }

    try {
      // Ensure directory exists
      const dir = path.substring(0, path.lastIndexOf('/'));
      if (dir) {
        this.pyodide.FS.mkdirTree(dir);
      }
      
      // Write file
      this.pyodide.FS.writeFile(path, content);
    } catch (error) {
      throw new Error(`Failed to write file ${path}: ${error}`);
    }
  }

  /**
   * Read a file from the Pyodide filesystem.
   */
  async readFile(path: string): Promise<string> {
    if (!this.initialized) {
      await this.initialize();
    }

    try {
      return this.pyodide.FS.readFile(path, { encoding: "utf8" });
    } catch (error) {
      throw new Error(`Failed to read file ${path}: ${error}`);
    }
  }
}

// Global persistent sandbox instance
let globalSandbox: PyodideSandbox | null = null;

// Main execution when called directly
// deno-lint-ignore-file no-explicit-any
if (import.meta.main) {
  // Initialize sandbox once on startup
  if (!globalSandbox) {
    console.error("Starting persistent Pyodide sandbox server...");
    globalSandbox = new PyodideSandbox();
    await globalSandbox.initialize();
    console.error("Pyodide sandbox server ready");
  }
  
  // Read commands from stdin in a loop
  const decoder = new TextDecoder();
  const reader = Deno.stdin.readable.getReader();
  
  while (true) {
    try {
      const { value, done } = await reader.read();
      if (done) {
        console.error("Stdin closed, exiting...");
        break;
      }
      
      if (value) {
        const lines = decoder.decode(value).trim().split('\n');
        for (const line of lines) {
          if (!line) continue;
          
          try {
            const request: ExecutionRequest = JSON.parse(line);
            
            // Handle different request types
            if (request.type === "writeFile" && request.path && request.content !== undefined) {
              // Handle file write request
              try {
                await globalSandbox.writeFile(request.path, request.content);
                const writeResult: ExecutionResult = {
                  success: true,
                  output: `File written to ${request.path}`,
                  error: null,
                  logs: [],
                };
                console.log(JSON.stringify(writeResult));
              } catch (error) {
                const errorResult: ExecutionResult = {
                  success: false,
                  output: "",
                  error: `Failed to write file: ${error}`,
                  logs: [],
                };
                console.log(JSON.stringify(errorResult));
              }
            } else {
              // Default to code execution (backward compatibility)
              if (!request.code) {
                const errorResult: ExecutionResult = {
                  success: false,
                  output: "",
                  error: "No code provided for execution",
                  logs: [],
                };
                console.log(JSON.stringify(errorResult));
                continue;
              }
              
              try {
                const result = await globalSandbox.execute(request);
                console.log(JSON.stringify(result));
              } catch (error) {
                const errorResult: ExecutionResult = {
                  success: false,
                  output: "",
                  error: `Execution failed: ${error}`,
                  logs: [],
                };
                console.log(JSON.stringify(errorResult));
              }
            }
          } catch (error) {
            // JSON parsing or other request processing errors
            const errorResult: ExecutionResult = {
              success: false,
              output: "",
              error: `Failed to process request: ${error}`,
              logs: [],
            };
            console.log(JSON.stringify(errorResult));
          }
        }
      }
    } catch (error) {
      // Log the error but don't break the loop unless it's a critical error
      console.error(`Server error: ${error}`);
      
      // Only break on critical errors that indicate the process should exit
      if (error instanceof Deno.errors.BrokenPipe || 
          error instanceof Deno.errors.ConnectionReset ||
          error.name === "BadResource") {
        console.error("Critical error detected, exiting...");
        break;
      }
      
      // For other errors, send an error response and continue
      try {
        const errorResult: ExecutionResult = {
          success: false,
          output: "",
          error: `Server error: ${error}`,
          logs: [],
        };
        console.log(JSON.stringify(errorResult));
      } catch (outputError) {
        console.error(`Failed to send error response: ${outputError}`);
        // If we can't even send an error response, the connection is likely broken
        break;
      }
    }
  }
  
  console.error("Pyodide sandbox server shutting down");
}

export { PyodideSandbox, type ExecutionRequest, type ExecutionResult };

File: src/wandb_mcp_server/mcp_tools/code_sandbox/pyodide_sandbox.py
"""
Pyodide local sandbox implementation using Deno for enhanced security.
"""

import asyncio
import asyncio.subprocess
import json
import subprocess
from pathlib import Path
from typing import Any, Dict, Optional

from wandb_mcp_server.utils import get_rich_logger
from .sandbox_utils import (
    SandboxError,
    validate_timeout,
    DEFAULT_TIMEOUT_SECONDS,
)

logger = get_rich_logger(__name__)


class PyodideSandbox:
    """Local Pyodide sandbox implementation using Deno for enhanced security."""
    
    # Class-level persistent process
    _shared_process = None
    _process_lock = asyncio.Lock()
    _initialized = False
    _initialization_error = None
    
    def __init__(self):
        self.available = self._check_deno_available()
        self._pyodide_script_path = Path(__file__).parent / "pyodide_sandbox.ts"
    
    async def __aenter__(self):
        """Context manager entry - ensure process is ready."""
        if self.available:
            await self.get_or_create_process(self._pyodide_script_path)
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit - process cleanup handled at shutdown."""
        # Don't cleanup process here as it's shared across executions
        return False
    
    def _check_deno_available(self) -> bool:
        """Check if Deno is available."""
        try:
            result = subprocess.run(
                ["deno", "--version"],
                capture_output=True,
                text=True,
                timeout=5
            )
            if result.returncode != 0:
                logger.warning(
                    "Deno is installed but returned an error. "
                    "Please ensure Deno is properly installed. "
                    "Visit https://deno.land/manual/getting_started/installation for installation instructions."
                )
                return False
            return True
        except FileNotFoundError:
            logger.warning(
                "Deno is not installed. Pyodide sandbox requires Deno for secure local execution. "
                "To install Deno:\n"
                "  - macOS/Linux: curl -fsSL https://deno.land/install.sh | sh\n"
                "  - Windows: irm https://deno.land/install.ps1 | iex\n"
                "  - Or visit: https://deno.land/manual/getting_started/installation"
            )
            return False
        except subprocess.TimeoutExpired:
            logger.warning("Deno check timed out. Please ensure Deno is properly installed.")
            return False
        except Exception as e:
            logger.warning(f"Error checking for Deno: {e}")
            return False
    
    @classmethod
    async def get_or_create_process(cls, script_path: Path):
        """Get or create the shared Pyodide process."""
        async with cls._process_lock:
            # Check if we have a cached initialization error
            if cls._initialization_error:
                raise cls._initialization_error
            
            if cls._shared_process is None or cls._shared_process.returncode is not None:
                logger.info("Starting persistent Pyodide sandbox process")
                try:
                    cls._shared_process = await asyncio.create_subprocess_exec(
                        'deno', 'run',
                        '--allow-net',  # For downloading Pyodide and packages
                        '--allow-read',  # To read local files
                        '--allow-write',  # To write output files and cache
                        '--allow-env',  # For environment variables
                        str(script_path),
                        stdin=asyncio.subprocess.PIPE,
                        stdout=asyncio.subprocess.PIPE,
                        stderr=asyncio.subprocess.PIPE
                    )
                    cls._initialized = True
                    
                    # Read initialization messages
                    try:
                        # Wait for "ready" message with timeout
                        ready_task = asyncio.create_task(cls._read_until_ready())
                        await asyncio.wait_for(ready_task, timeout=60)
                        logger.info("Pyodide sandbox process initialized successfully")
                        cls._initialization_error = None  # Clear any previous error
                    except asyncio.TimeoutError:
                        logger.error("Timeout waiting for Pyodide to initialize")
                        if cls._shared_process:
                            cls._shared_process.kill()
                            await cls._shared_process.wait()
                        cls._shared_process = None
                        error = SandboxError(
                            "Failed to initialize Pyodide sandbox. "
                            "This may be due to network issues downloading Pyodide. "
                            "Please check your internet connection and try again."
                        )
                        cls._initialization_error = error
                        raise error
                except FileNotFoundError:
                    error = SandboxError(
                        "Deno executable not found. Please install Deno to use the Pyodide sandbox.\n"
                        "Installation instructions: https://deno.land/manual/getting_started/installation"
                    )
                    cls._initialization_error = error
                    raise error
                except Exception as e:
                    error = SandboxError(f"Failed to start Pyodide sandbox process: {e}")
                    cls._initialization_error = error
                    raise error
                    
            return cls._shared_process
    
    @classmethod
    async def _read_until_ready(cls):
        """Read stderr until we see the ready message."""
        if not cls._shared_process:
            return
            
        while True:
            line = await cls._shared_process.stderr.readline()
            if not line:
                break
            line_str = line.decode('utf-8').strip()
            logger.debug(f"Pyodide init: {line_str}")
            if "Pyodide sandbox server ready" in line_str:
                break
            elif "Initializing Pyodide..." in line_str:
                logger.info("Downloading and initializing Pyodide (this may take a minute on first run)...")
    
    async def execute_code(self, code: str, timeout: int = DEFAULT_TIMEOUT_SECONDS) -> Dict[str, Any]:
        """Execute Python code using our persistent Pyodide process."""
        # Validate timeout
        timeout = validate_timeout(timeout)
        
        if not self.available:
            raise SandboxError(
                "Deno is not available for Pyodide execution.\n"
                "To install Deno:\n"
                "  - macOS/Linux: curl -fsSL https://deno.land/install.sh | sh\n"
                "  - Windows: irm https://deno.land/install.ps1 | iex\n"
                "  - Or visit: https://deno.land/manual/getting_started/installation"
            )
        
        if not self._pyodide_script_path.exists():
            raise SandboxError(f"Pyodide script not found at {self._pyodide_script_path}")
        
        try:
            # Get or create the persistent process
            process = await self.get_or_create_process(self._pyodide_script_path)
            
            # Create execution request
            request = {
                "type": "execute",
                "code": code,
                "timeout": timeout
            }
            
            # Send request as a single line JSON
            request_json = json.dumps(request) + '\n'
            process.stdin.write(request_json.encode())
            await process.stdin.drain()
            
            # Read response with timeout
            try:
                response_line = await asyncio.wait_for(
                    process.stdout.readline(),
                    timeout=timeout + 5  # Give extra time for process overhead
                )
            except asyncio.TimeoutError:
                # Don't kill the process on timeout, just return error
                return {
                    "success": False,
                    "output": "",
                    "error": f"Execution timed out after {timeout} seconds",
                    "logs": [],
                }
            
            # Parse the response
            try:
                if response_line:
                    result = json.loads(response_line.decode('utf-8'))
                    return result
                else:
                    return {
                        "success": False,
                        "output": "",
                        "error": "No response from Pyodide process",
                        "logs": [],
                    }
                    
            except json.JSONDecodeError as e:
                return {
                    "success": False,
                    "output": "",
                    "error": f"Failed to parse response: {e}",
                    "logs": [],
                }
            
        except SandboxError:
            raise
        except Exception as e:
            return {
                "success": False,
                "output": "",
                "error": f"Pyodide execution failed: {str(e)}",
                "logs": [],
            }
    
    async def writeFile(self, path: str, content: str) -> None:
        """Write a file to the Pyodide sandbox using persistent process."""
        if not self.available:
            raise SandboxError(
                "Deno is not available for Pyodide execution. "
                "Install Deno with: curl -fsSL https://deno.land/install.sh | sh"
            )
        
        try:
            # Get or create the persistent process
            process = await self.get_or_create_process(self._pyodide_script_path)
            
            # Create file write request
            request = {
                "type": "writeFile",
                "path": path,
                "content": content
            }
            
            # Send request as a single line JSON
            request_json = json.dumps(request) + '\n'
            process.stdin.write(request_json.encode())
            await process.stdin.drain()
            
            # Read response with timeout
            try:
                response_line = await asyncio.wait_for(
                    process.stdout.readline(),
                    timeout=10
                )
            except asyncio.TimeoutError:
                raise SandboxError(f"Timeout writing file to {path}")
            
            # Parse the response
            if response_line:
                result = json.loads(response_line.decode('utf-8'))
                if not result.get("success", False):
                    raise SandboxError(f"Failed to write file: {result.get('error', 'Unknown error')}")
            else:
                raise SandboxError("No response from Pyodide process")
                
        except Exception as e:
            if isinstance(e, SandboxError):
                raise
            raise SandboxError(f"Failed to write file to Pyodide: {str(e)}")
    
    @classmethod
    async def initialize_early(cls):
        """Initialize Pyodide early to download packages during server startup."""
        sandbox = cls()
        if not sandbox.available:
            logger.info("Deno not available, skipping Pyodide early initialization")
            return False
            
        logger.info("Starting early Pyodide initialization to download packages...")
        try:
            # Create the process and send a simple execute request to trigger initialization
            await sandbox.get_or_create_process(sandbox._pyodide_script_path)
            
            # Execute a simple command to trigger package loading
            result = await sandbox.execute("print('Pyodide initialized successfully')", timeout=60)
            
            if result.get("success"):
                logger.info("Pyodide packages downloaded and ready for use")
                return True
            else:
                logger.warning(f"Pyodide initialization completed with error: {result.get('error')}")
                return False
                
        except Exception as e:
            logger.warning(f"Failed to initialize Pyodide early: {e}")
            return False
    
    @classmethod
    async def cleanup_shared_process(cls):
        """Explicitly close the shared Pyodide process (for cleanup)."""
        async with cls._process_lock:
            if cls._shared_process is not None:
                try:
                    # First try graceful termination
                    cls._shared_process.terminate()
                    try:
                        # Wait for process to exit with timeout
                        await asyncio.wait_for(cls._shared_process.wait(), timeout=5.0)
                        logger.info("Pyodide process terminated gracefully")
                    except asyncio.TimeoutError:
                        # Force kill if termination times out
                        logger.warning("Pyodide process did not terminate gracefully, forcing kill")
                        cls._shared_process.kill()
                        await cls._shared_process.wait()
                        logger.info("Pyodide process killed forcefully")
                except Exception as e:
                    logger.error(f"Error closing shared Pyodide process: {e}")
                finally:
                    cls._shared_process = None
                    cls._initialized = False
                    cls._initialization_error = None